lib/Driver/ToolChains.cpp

```diff
@@ -526,6 +353,186 @@ ToolChain::constructInvocation(const CompileJobAction &job,
   return II;
 }
 
+const char *ToolChain::JobContext::computeFrontendModeForCompile() const {
```

Comment:
This one still makes me uncomfortable. Why does JobContext know about fontend flags? The same struct is used for all the different kinds of jobs, but this one's only relevant for compile jobs.

Reply:
Interesting. This function does have problems: it's doing two things, filtering out Immediate and REPL compiles, then mapping primary output type to compilation mode arguments. I agree that it's misplaced. Where does the knowledge of modes belong? How about in CompileJobAction? Would that be better?

Reply:
(planning to talk about this in person today)

Reply:
(We decided that it's okay to leave these on JobContext as long as they have "Frontend" in the name.)